components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/AbstractAuthenticationHandler.java

```diff
@@ -0,0 +1,152 @@
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.auth.service.handler;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.http.HttpHeaders;
+import org.wso2.carbon.identity.auth.service.AuthenticationContext;
+import org.wso2.carbon.identity.auth.service.AuthenticationResult;
+import org.wso2.carbon.identity.auth.service.exception.AuthClientException;
+import org.wso2.carbon.identity.auth.service.exception.AuthServerException;
+import org.wso2.carbon.identity.auth.service.exception.AuthenticationFailException;
+import org.wso2.carbon.identity.common.base.handler.AbstractMessageHandler;
+import org.wso2.carbon.identity.common.base.message.MessageContext;
+import org.wso2.carbon.identity.mgt.RealmService;
+
+import java.util.Locale;
+
+/**
+ * This is the abstract class for custom authentication handlers.
+ *
+ * The custom handlers should implement the doAuthenticate() method and optionally the postAuthenticate() method.
+ *
+ */
+public abstract class AbstractAuthenticationHandler extends AbstractMessageHandler implements AuthenticationHandler {
+
+    private RealmService realmService;
+
+    @Override
+    public final AuthenticationResult authenticate(MessageContext messageContext)
+            throws AuthServerException, AuthenticationFailException, AuthClientException {
+
+        AuthenticationResult authenticationResult = this.doAuthenticate(messageContext);
+        postAuthenticate(messageContext, authenticationResult);
+
+        return authenticationResult;
+
+    }
+
+    public boolean isEnabled(MessageContext messageContext) {
+        if (messageContext instanceof AuthenticationContext) {
+            AuthenticationContext authenticationContext = (AuthenticationContext) messageContext;
+            if (authenticationContext.getAuthenticationRequest() != null) {
+                String authorizationHeader = authenticationContext.getAuthenticationRequest().
+                        getHeader(HttpHeaders.AUTHORIZATION);
+                if (StringUtils.isNotEmpty(authorizationHeader) && authorizationHeader
+                        .startsWith(getAuthorizationHeaderType())) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Returns the domain name from the user string.
+     * @param userName
+     * @return
+     */
+    protected ImmutablePair<String, String> decodeTenantDomainAndUserName(String userName) {
+        String domainName = null;
+        if(userName != null && userName.contains("/") ) {
+            int lastAt = userName.lastIndexOf('/');
+            domainName = userName.substring(0,lastAt);
+            userName = userName.substring(lastAt + 1);
+        }
+        if(domainName != null) {
+            domainName = domainName.toLowerCase(Locale.getDefault());
+        }
+        if(domainName == null) {
+            domainName = "PRIMARY";
```

Comment:
We cannot guarantee that the primary domain name will be "PRIMARY". If needed we can either use `identityStore.getPrimaryDomain()`.

If the domain is not available, is it reasonable to assume the domain is primary ? If not, we can just pass `null` as the domain, so the authentication will happen across all domains.